streams/src/main/java/org/apache/kafka/streams/state/internals/ListValueStore.java

```diff
@@ -0,0 +1,147 @@
+import org.apache.kafka.common.serialization.Serdes;
+import org.apache.kafka.common.serialization.Serde;
+
+/**
+ * A wrapper key-value store that serializes the record values bytes as a list.
+ * As a result put calls would be interpreted as a get-append-put to the underlying RocksDB store.
+ * Range iterators would also flatten the value lists and return the values one-by-one.
+ *
+ * This store is used for cases where we do not want to de-duplicate values of the same keys but want to retain all such values.
+ */
+@SuppressWarnings("unchecked")
+public class ListValueStore
+    extends WrappedStateStore<KeyValueStore<Bytes, byte[]>, Bytes, byte[]>
+    implements KeyValueStore<Bytes, byte[]> {
+
+    static private final Serde<List<byte[]>> LIST_SERDE = Serdes.ListSerde(ArrayList.class, Serdes.ByteArray());
+
+    ListValueStore(final KeyValueStore<Bytes, byte[]> bytesStore) {
+        super(bytesStore);
+    }
+
+    @Override
+    public void put(final Bytes key, final byte[] value) {
+        // if the value is null we can skip the get and blind delete
+        if (value == null) {
+            wrapped().put(key, null);
+        } else {
+            final byte[] oldValue = wrapped().get(key);
+
+            if (oldValue == null) {
+                wrapped().put(key, LIST_SERDE.serializer().serialize(null, Collections.singletonList(value)));
+            } else {
+                final List<byte[]> list = LIST_SERDE.deserializer().deserialize(null, oldValue);
+                list.add(value);
+
+                wrapped().put(key, LIST_SERDE.serializer().serialize(null, list));
+            }
+        }
+    }
+
+    @Override
+    public byte[] putIfAbsent(final Bytes key, final byte[] value) {
+        throw new UnsupportedOperationException("putIfAbsent not supported");
+    }
+
+    @Override
+    public void putAll(final List<KeyValue<Bytes, byte[]>> entries) {
+        throw new UnsupportedOperationException("putAll not supported");
+    }
+
+    @Override
+    public byte[] delete(final Bytes key) {
+        // we intentionally disable delete calls since the returned bytes would
+        // represent a list, not a single value; we need to have a new API for delete if we do need it
+        throw new UnsupportedOperationException("delete not supported");
+    }
+
+    @Override
+    public byte[] get(final Bytes key) {
+        return wrapped().get(key);
+    }
+
+    @Override
+    public KeyValueIterator<Bytes, byte[]> range(final Bytes from, final Bytes to) {
+        throw new UnsupportedOperationException("range not supported");
+    }
+
+    @Override
+    public KeyValueIterator<Bytes, byte[]> all() {
+        return new WrappedStoreIterator(wrapped().all());
+    }
+
+    @Override
+    public long approximateNumEntries() {
+        return wrapped().approximateNumEntries();
+    }
+
+    private static class WrappedStoreIterator extends AbstractIterator<KeyValue<Bytes, byte[]>>
```

Comment:
Should we call this class `ValueListIterator` (or similar) -- "Wrapped" is not very descriptive.

Reply:
Ack.